### PR TITLE
Robustify Logging

### PR DESCRIPTION
**Pull request by alecholmez**

### CODE REVIEW

File: pkg/log/log.go [hunk 1]
```diff
@@ -15,11 +15,49 @@
 // Package log provides a logging interface for use in this library.
 package log
 
+import "strings"
+
+// Level is an exposed logger level type that users can select
+type Level string
+
+const (
```
Comment (alecholmez) [inline pkg/log/log.go]: I'm not sure if these belong in here.... would like some input

File: pkg/log/log.go [hunk 2]
```diff
@@ -15,11 +15,49 @@
...
+	// INFO Log Level (little less log pollution from xDS)
+	INFO Level = "info"
+	// WARN Log Level (show warnigs and potential hazards)
+	WARN Level = "warn"
+	// ERROR Log Level (only log failures)
+	ERROR Level = "error"
+)
+
+// ParseLevel will accept a string and return a logging level
+func ParseLevel(l string) Level {
```
Comment (alecholmez) [inline pkg/log/log.go]: Same thing here, I'm not sure if this belongs in this package, it's more of an implementation thing

File: pkg/test/main/main.go [hunk 3]
```diff
@@ -150,7 +154,7 @@ func main() {
 		}
 
 		als.Dump(func(s string) {
-			if debug {
+			if level == l.DEBUG {
```
Comment (joshua-rutherford) [inline pkg/test/main/main.go]: Why aren't we allowing the leveled logger handle this check?

File: pkg/test/main/main.go [hunk 4]
```diff
@@ -219,13 +223,28 @@ func callEcho() (int, int) {
 
 type logger struct{}
 
+func (logger logger) Debugf(format string, args ...interface{}) {
+	if level == l.DEBUG {
```
Comment (joshua-rutherford) [inline pkg/test/main/main.go]: So I don't here's the flaw in this approach.  You need to make the log levels numeric so you can say:

```
if level >= l.DEBUG {
...
}
```

Not terribly important on your debug level, this approach means info messages aren't logged when debug is set.

File: pkg/test/main/main.go [hunk 5]
```diff
@@ -58,7 +61,7 @@ var (
 )
 
 func init() {
-	flag.BoolVar(&debug, "debug", false, "Use debug logging")
+	flag.StringVar(&loggerLevel, "debug", "debug", "Use debug logging")
```
[the remark this answers is not in the document]
Comment (alecholmez) [reply pkg/test/main/main.go]: Good point, I'm going to revert to what they add for the integration test

File: pkg/test/main/main.go (hunk 3 above)
Comment (alecholmez) [reply pkg/test/main/main.go]: Reverting back to their pre-existing logic

File: pkg/test/main/main.go (hunk 4 above)
Comment (alecholmez) [reply pkg/test/main/main.go]: Valid, this actually isn't critical anyways so I'm going to revert back to their previous logic. It makes more sense for whoever is implementing the logging interface to deal with this

File: go.mod [hunk 6]
```diff
@@ -4,10 +4,11 @@ go 1.12
...
 require (
 	github.com/census-instrumentation/opencensus-proto v0.2.1
+	github.com/cncf/udpa/go v0.0.0-20191209042840-269d4d468f6f
 	github.com/envoyproxy/protoc-gen-validate v0.1.0
 	github.com/golang/protobuf v1.3.2
 	github.com/google/go-cmp v0.2.0
 	github.com/prometheus/client_model v0.0.0-20190812154241-14fe0d1b01d4
 	google.golang.org/genproto v0.0.0-20190819201941-24fa4b261c55
-	google.golang.org/grpc v1.23.0
+	google.golang.org/grpc v1.25.1
```
[the remark this answers is not in the document]
Comment (alecholmez) [reply go.mod]: That version got bumped when I ran the build, I'll see what the upstream says about upgrading

**Pull request merged**